physicalplan/binary/table.go

```diff
@@ -94,20 +94,91 @@ func (t *table) execBinaryOperation(lhs model.StepVector, rhs model.StepVector)
+// if it is a vector comparison operation.
+type operation func(lhs, rhs float64, scalarSideLeft bool) (float64, bool)
 
 var operations = map[string]operation{
-	"+": func(lhs float64, rhs float64) float64 { return lhs + rhs },
-	"-": func(lhs float64, rhs float64) float64 { return lhs - rhs },
-	"*": func(lhs float64, rhs float64) float64 { return lhs * rhs },
-	"/": func(lhs float64, rhs float64) float64 { return lhs / rhs },
+	"+":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return lhs + rhs, true },
+	"-":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return lhs - rhs, true },
+	"*":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return lhs * rhs, true },
+	"/":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return lhs / rhs, true },
+	"^":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return math.Pow(lhs, rhs), true },
+	"%":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return math.Mod(lhs, rhs), true },
+	"==":    func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs == rhs), true },
+	"!=":    func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs != rhs), true },
+	">":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs > rhs), true },
+	"<":     func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs < rhs), true },
+	">=":    func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs >= rhs), true },
+	"<=":    func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return btos(lhs <= rhs), true },
+	"atan2": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) { return math.Atan2(lhs, rhs), true },
 }
 
-func newOperation(expr parser.ItemType) (operation, error) {
+// For vector, those operations are handled differently to check whether to keep
+// the value or not. https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L2229
+var vectorBinaryOperations = map[string]operation{
+	"==": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
+			res = rhs
+		}
+		return res, lhs == rhs
+	},
+	"!=": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
+			res = rhs
+		}
+		return res, lhs != rhs
+	},
+	">": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
+			res = rhs
+		}
+		return res, lhs > rhs
+	},
+	"<": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
+			res = rhs
+		}
+		return res, lhs < rhs
+	},
+	">=": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
+			res = rhs
+		}
+		return res, lhs >= rhs
+	},
+	"<=": func(lhs, rhs float64, scalarSideLeft bool) (float64, bool) {
+		res := lhs
+		if scalarSideLeft {
```

Comment:
Hm, I don't think the if statement should be here inside. When the scalar is on the right side, we will get a lot of branch mispredictions in the CPU which will break instruction pipelining. We might want to swap the arguments somewhere earlier instead.

Reply:
I agree. I just updated a version which avoids the condition check. Does it look better now?

Reply:
Yes, looks good